File: scripts/ci/assert-trusted-target-inventory.py
#!/usr/bin/env python3
"""Validate P0.0 trusted target-inventory fixture coverage without live claims.

This checker is local/static fixture evidence only. It proves that the checked-in
trusted target-inventory schema and baseline GOOD/BAD fixtures exercise the
requirement that Buck2 build/test targets come from trusted dev/controller state
before candidate checkout. It does not claim live cloud-ci execution, protected
branch authority, P0.0 green, or Phase-0 completion.
"""

from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path
from typing import Any

REQUIRED_INVENTORY_FIELDS = [
    "candidate_sha",
    "claim_boundary",
    "inventory_source",
    "captured_before_candidate_checkout",
    "candidate_checkout_after_inventory",
    "no_candidate_authored_discovery",
    "build_targets",
    "test_targets",
    "expected_verdict",
    "expected_violations",
    "fixture_id",
    "source_test",
]

TRUSTED_SOURCE = "trusted_dev_or_controller_state"
CANDIDATE_SOURCE = "candidate_pr_bytes"
TRUSTED_TARGET_VIOLATIONS = {
    "target_inventory_not_trusted",
    "inventory_not_captured_before_candidate_checkout",
    "candidate_can_author_target_inventory",
    "empty_required_targets",
    "malformed_buck2_target",
    "green_claim_boundary_without_live_authority",
}
SHA_RE = re.compile(r"^[0-9a-fA-F]{40}$")
TARGET_RE = re.compile(r"^(?:[A-Za-z0-9_.-]+)?//[A-Za-z0-9_./-]*:[A-Za-z0-9_+=.,@~-]+$")

DEFAULT_SCHEMA = "specs/phase0-trusted-target-inventory-schema.json"
DEFAULT_GOOD_FIXTURE = "specs/fixtures/phase0-ci-enforcement-baseline/tc-0.0.1a-good-trusted-target-inventory.json"
DEFAULT_BAD_FIXTURE = "specs/fixtures/phase0-ci-enforcement-baseline/tc-0.0.1a-bad-candidate-sourced-target-inventory.json"


def load_json(path: str | Path) -> dict[str, Any]:
    with Path(path).open() as fh:
        data = json.load(fh)
    if not isinstance(data, dict):
        raise TypeError(f"{path}: expected JSON object")
    return data


def string_list(value: Any) -> list[str]:
    if not isinstance(value, list):
        return []
    return [item for item in value if isinstance(item, str)]


def require(condition: bool, failures: list[str], message: str) -> None:
    if not condition:
        failures.append(message)


def validate_schema(schema: dict[str, Any], failures: list[str]) -> dict[str, Any]:
    required = string_list(schema.get("required"))
    for field in REQUIRED_INVENTORY_FIELDS:
        require(field in required, failures, f"schema.required missing {field}")
    require(schema.get("additionalProperties") is False, failures, "schema.additionalProperties must be false")

    properties = schema.get("properties")
    if not isinstance(properties, dict):
        failures.append("schema.properties must be an object")
        properties = {}

    inventory_source = properties.get("inventory_source") if isinstance(properties.get("inventory_source"), dict) else {}
    for field in REQUIRED_INVENTORY_FIELDS:
        require(field in properties, failures, f"schema.properties missing {field}")

    source_enum = set(string_list(inventory_source.get("enum")))
    require(source_enum == {TRUSTED_SOURCE, CANDIDATE_SOURCE}, failures, "schema.inventory_source enum must be exactly trusted_dev_or_controller_state and candidate_pr_bytes")

    verdict = properties.get("expected_verdict") if isinstance(properties.get("expected_verdict"), dict) else {}
    verdict_enum = set(string_list(verdict.get("enum")))
    require(verdict_enum == {"GREEN", "RED"}, failures, "schema.expected_verdict enum must be exactly GREEN and RED")

    claim_boundary = properties.get("claim_boundary") if isinstance(properties.get("claim_boundary"), dict) else {}
    claim_required = set(string_list(claim_boundary.get("required")))
    require({"p0_0_green", "phase0_complete"}.issubset(claim_required), failures, "schema.claim_boundary must require p0_0_green and phase0_complete")
    require(claim_boundary.get("additionalProperties") is False, failures, "schema.claim_boundary.additionalProperties must be false")

    for field in ["build_targets", "test_targets"]:
        target_field = properties.get(field) if isinstance(properties.get(field), dict) else {}
        require(target_field.get("minItems") == 1, failures, f"schema.{field}.minItems must be 1")

    return {
        "required_fields": required,
        "inventory_source_values": sorted(source_enum),
        "expected_verdict_values": sorted(verdict_enum),
    }


def target_list(value: Any) -> list[str]:
    if not isinstance(value, list):
        return []
    return [item for item in value if isinstance(item, str)]


def malformed_targets(targets: list[str]) -> list[str]:
    return [target for target in targets if not TARGET_RE.match(target)]


def inventory_fixture_shape_failures(fixture: dict[str, Any]) -> list[str]:
    failures: list[str] = []
    unknown_fields = sorted(set(fixture) - set(REQUIRED_INVENTORY_FIELDS))
    if unknown_fields:
        failures.append(f"unexpected top-level fields: {unknown_fields}")
    candidate_sha = fixture.get("candidate_sha")
    if not isinstance(candidate_sha, str) or not SHA_RE.match(candidate_sha):
        failures.append("candidate_sha must be a 40-character hexadecimal SHA")
    return failures


def inventory_fixture_violations(fixture: dict[str, Any]) -> list[str]:
    violations: list[str] = []
    if fixture.get("inventory_source") != TRUSTED_SOURCE:
        violations.append("target_inventory_not_trusted")
    if fixture.get("captured_before_candidate_checkout") is not True or fixture.get("candidate_checkout_after_inventory") is not True:
        violations.append("inventory_not_captured_before_candidate_checkout")
    if fixture.get("no_candidate_authored_discovery") is not True:
        violations.append("candidate_can_author_target_inventory")

    build_targets = target_list(fixture.get("build_targets"))
    test_targets = target_list(fixture.get("test_targets"))
    if not build_targets or not test_targets:
        violations.append("empty_required_targets")
    if malformed_targets(build_targets + test_targets):
        violations.append("malformed_buck2_target")

    claim_boundary = fixture.get("claim_boundary")
    if not isinstance(claim_boundary, dict) or claim_boundary.get("p0_0_green") is not False or claim_boundary.get("phase0_complete") is not False:
        violations.append("green_claim_boundary_without_live_authority")
    return violations


def validate_fixture(path: str, expected_verdict: str, failures: list[str]) -> dict[str, Any]:
    fixture = load_json(path)
    fixture_id = fixture.get("fixture_id", path)
    require(fixture.get("expected_verdict") == expected_verdict, failures, f"{fixture_id}: expected_verdict must be {expected_verdict}")
    shape_failures = inventory_fixture_shape_failures(fixture)
    for shape_failure in shape_failures:
        failures.append(f"{fixture_id}: {shape_failure}")
    observed_violations = inventory_fixture_violations(fixture)
    expected_violations = set(string_list(fixture.get("expected_violations")))
    if expected_verdict == "GREEN":
        require(not observed_violations, failures, f"{fixture_id}: GOOD trusted-target inventory has violations {observed_violations}")
        require(not expected_violations, failures, f"{fixture_id}: GOOD fixture must not list expected violations")
    else:
        require(bool(observed_violations), failures, f"{fixture_id}: RED trusted-target inventory must produce violations")
        require(TRUSTED_TARGET_VIOLATIONS.issubset(expected_violations), failures, f"{fixture_id}: RED fixture expected_violations must include all trusted-target violation classes")
        require(set(observed_violations).issubset(expected_violations), failures, f"{fixture_id}: observed trusted-target violations not listed in expected_violations")
    return {
        "fixture_id": fixture_id,
        "expected_verdict": expected_verdict,
        "shape_failures": shape_failures,
        "observed_trusted_target_violations": observed_violations,
        "malformed_targets": malformed_targets(target_list(fixture.get("build_targets")) + target_list(fixture.get("test_targets"))),
        "trusted_inventory": not shape_failures and not observed_violations,
    }


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--schema", default=DEFAULT_SCHEMA)
    parser.add_argument("--good-fixture", default=DEFAULT_GOOD_FIXTURE)
    parser.add_argument("--bad-fixture", default=DEFAULT_BAD_FIXTURE)
    parser.add_argument("--json", action="store_true")
    args = parser.parse_args()

    failures: list[str] = []
    schema_summary = validate_schema(load_json(args.schema), failures)
    fixture_results = [
        validate_fixture(args.good_fixture, "GREEN", failures),
        validate_fixture(args.bad_fixture, "RED", failures),
    ]

    result = {
        "authority_boundary": "trusted target-inventory fixture evidence only; this checker never claims live cloud-ci execution or protected-branch authority",
        "schema": args.schema,
        "required_inventory_fields": REQUIRED_INVENTORY_FIELDS,
        "required_trusted_target_violations": sorted(TRUSTED_TARGET_VIOLATIONS),
        "schema_summary": schema_summary,
        "fixture_results": fixture_results,
        "local_fixture_contract_proven": not failures,
        "candidate_pr_bytes_are_data_only_locally_proven": not failures,
        "trusted_target_inventory_live_authority_proven": False,
        "trusted_controller_inventory_live": False,
        "live_required_context_execution_proven": False,
        "p0_0_green": False,
        "phase0_complete": False,
        "verdict": "PASS" if not failures else "FAIL",
        "failures": failures,
    }
    rendered = json.dumps(result, sort_keys=True)
    if args.json or not failures:
        print(rendered)
    else:
        print(rendered, file=sys.stderr)
    return 0 if not failures else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/ci/assert-trusted-target-inventory.rs
//! Validate P0.0 trusted target-inventory fixture coverage without live claims.
//!
//! This checker is local/static fixture evidence only. It proves that the
//! checked-in trusted target-inventory schema and baseline GOOD/BAD fixtures
//! exercise the requirement that Buck2 build/test targets come from trusted
//! dev/controller state before candidate checkout. It does not claim live
//! cloud-ci execution, protected branch authority, P0.0 green, or Phase-0
//! completion.

#[allow(dead_code)]
#[path = "assert-result-bundle-output.rs"]
mod json_support;

pub use json_support::Json;
use std::collections::{BTreeMap, BTreeSet};
use std::env;
use std::fs;

pub const REQUIRED_INVENTORY_FIELDS: [&str; 12] = [
    "candidate_sha",
    "claim_boundary",
    "inventory_source",
    "captured_before_candidate_checkout",
    "candidate_checkout_after_inventory",
    "no_candidate_authored_discovery",
    "build_targets",
    "test_targets",
    "expected_verdict",
    "expected_violations",
    "fixture_id",
    "source_test",
];

const TRUSTED_SOURCE: &str = "trusted_dev_or_controller_state";
const CANDIDATE_SOURCE: &str = "candidate_pr_bytes";
const TRUSTED_TARGET_VIOLATIONS: [&str; 6] = [
    "candidate_can_author_target_inventory",
    "empty_required_targets",
    "green_claim_boundary_without_live_authority",
    "inventory_not_captured_before_candidate_checkout",
    "malformed_buck2_target",
    "target_inventory_not_trusted",
];

const DEFAULT_SCHEMA: &str = "specs/phase0-trusted-target-inventory-schema.json";
const DEFAULT_GOOD_FIXTURE: &str =
    "specs/fixtures/phase0-ci-enforcement-baseline/tc-0.0.1a-good-trusted-target-inventory.json";
const DEFAULT_BAD_FIXTURE: &str = "specs/fixtures/phase0-ci-enforcement-baseline/tc-0.0.1a-bad-candidate-sourced-target-inventory.json";
const AUTHORITY_BOUNDARY: &str = "trusted target-inventory fixture evidence only; this checker never claims live cloud-ci execution or protected-branch authority";

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct SchemaSummary {
    pub required_fields: Vec<String>,
    pub inventory_source_values: Vec<String>,
    pub expected_verdict_values: Vec<String>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct FixtureValidation {
    pub fixture_id: String,
    pub expected_verdict: String,
    pub shape_failures: Vec<String>,
    pub observed_trusted_target_violations: Vec<String>,
    pub malformed_targets: Vec<String>,
    pub trusted_inventory: bool,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Report {
    pub authority_boundary: String,
    pub schema: String,
    pub required_inventory_fields: Vec<String>,
    pub required_trusted_target_violations: Vec<String>,
    pub schema_summary: SchemaSummary,
    pub fixture_results: Vec<FixtureValidation>,
    pub local_fixture_contract_proven: bool,
    pub candidate_pr_bytes_are_data_only_locally_proven: bool,
    pub trusted_target_inventory_live_authority_proven: bool,
    pub trusted_controller_inventory_live: bool,
    pub live_required_context_execution_proven: bool,
    pub p0_0_green: bool,
    pub phase0_complete: bool,
    pub verdict: String,
    pub failures: Vec<String>,
}

pub fn parse_json(text: &str) -> Result<Json, String> {
    json_support::parse_json(text)
}

pub fn load_json(path: &str) -> Result<Json, String> {
    let text = fs::read_to_string(path).map_err(|error| format!("read {path} failed: {error}"))?;
    parse_json(&text).map_err(|error| format!("parse {path} failed: {error}"))
}

fn object_field<'a>(object: &'a BTreeMap<String, Json>, key: &str) -> Option<&'a Json> {
    object.get(key)
}

fn string_field(object: &BTreeMap<String, Json>, key: &str) -> Option<String> {
    object_field(object, key)
        .and_then(Json::as_str)
        .map(str::to_string)
}

fn bool_field(object: &BTreeMap<String, Json>, key: &str) -> Option<bool> {
    object_field(object, key).and_then(Json::as_bool)
}

fn string_list(value: Option<&Json>) -> Vec<String> {
    value
        .and_then(Json::as_array)
        .map(|items| {
            items
                .iter()
                .filter_map(Json::as_str)
                .map(str::to_string)
                .collect()
        })
        .unwrap_or_default()
}

fn enum_values(properties: &BTreeMap<String, Json>, field: &str) -> BTreeSet<String> {
    object_field(properties, field)
        .and_then(Json::as_object)
        .map(|object| string_list(object_field(object, "enum")))
        .unwrap_or_default()
        .into_iter()
        .collect()
}

fn set_from(values: &[&str]) -> BTreeSet<String> {
    values.iter().map(|value| (*value).to_string()).collect()
}

fn sorted_unique(values: Vec<String>) -> Vec<String> {
    values
        .into_iter()
        .collect::<BTreeSet<_>>()
        .into_iter()
        .collect()
}

fn require(condition: bool, failures: &mut Vec<String>, message: String) {
    if !condition {
        failures.push(message);
    }
}

fn is_hex_sha(value: &str) -> bool {
    value.len() == 40 && value.bytes().all(|byte| byte.is_ascii_hexdigit())
}

fn allowed_chars(value: &str, allowed: fn(u8) -> bool) -> bool {
    value.bytes().all(allowed)
}

fn prefix_char(byte: u8) -> bool {
    byte.is_ascii_alphanumeric() || matches!(byte, b'_' | b'.' | b'-')
}

fn package_char(byte: u8) -> bool {
    byte.is_ascii_alphanumeric() || matches!(byte, b'_' | b'.' | b'/' | b'-')
}

fn target_name_char(byte: u8) -> bool {
    byte.is_ascii_alphanumeric()
        || matches!(byte, b'_' | b'+' | b'=' | b'.' | b',' | b'@' | b'~' | b'-')
}

fn is_buck2_target(value: &str) -> bool {
    let Some(double_slash_index) = value.find("//") else {
        return false;
    };
    let prefix = &value[..double_slash_index];
    if !prefix.is_empty() && !allowed_chars(prefix, prefix_char) {
        return false;
    }
    let after_slash = &value[double_slash_index + 2..];
    let Some(colon_index) = after_slash.rfind(':') else {
        return false;
    };
    let package = &after_slash[..colon_index];
    let name = &after_slash[colon_index + 1..];
    !name.is_empty()
        && allowed_chars(package, package_char)
        && allowed_chars(name, target_name_char)
}

fn target_list(value: Option<&Json>) -> Vec<String> {
    string_list(value)
}

fn malformed_targets(targets: &[String]) -> Vec<String> {
    targets
        .iter()
        .filter(|target| !is_buck2_target(target))
        .cloned()
        .collect()
}

fn validate_schema(schema: &Json, failures: &mut Vec<String>) -> SchemaSummary {
    let Some(object) = schema.as_object() else {
        failures.push("schema must be a JSON object".to_string());
        return SchemaSummary {
            required_fields: Vec::new(),
            inventory_source_values: Vec::new(),
            expected_verdict_values: Vec::new(),
        };
    };
    let required = string_list(object_field(object, "required"));
    for field in REQUIRED_INVENTORY_FIELDS {
        require(
            required.iter().any(|item| item == field),
            failures,
            format!("schema.required missing {field}"),
        );
    }
    require(
        bool_field(object, "additionalProperties") == Some(false),
        failures,
        "schema.additionalProperties must be false".to_string(),
    );

    let properties = object_field(object, "properties").and_then(Json::as_object);
    if properties.is_none() {
        failures.push("schema.properties must be an object".to_string());
    }
    let empty = BTreeMap::new();
    let properties = properties.unwrap_or(&empty);
    for field in REQUIRED_INVENTORY_FIELDS {
        require(
            properties.contains_key(field),
            failures,
            format!("schema.properties missing {field}"),
        );
    }

    let source_enum = enum_values(properties, "inventory_source");
    require(
        source_enum == set_from(&[TRUSTED_SOURCE, CANDIDATE_SOURCE]),
        failures,
        "schema.inventory_source enum must be exactly trusted_dev_or_controller_state and candidate_pr_bytes".to_string(),
    );
    let verdict_enum = enum_values(properties, "expected_verdict");
    require(
        verdict_enum == set_from(&["GREEN", "RED"]),
        failures,
        "schema.expected_verdict enum must be exactly GREEN and RED".to_string(),
    );

    let claim_boundary = object_field(properties, "claim_boundary").and_then(Json::as_object);
    let claim_boundary_empty = BTreeMap::new();
    let claim_boundary = claim_boundary.unwrap_or(&claim_boundary_empty);
    let claim_required = string_list(object_field(claim_boundary, "required"))
        .into_iter()
        .collect::<BTreeSet<_>>();
    require(
        set_from(&["p0_0_green", "phase0_complete"]).is_subset(&claim_required),
        failures,
        "schema.claim_boundary must require p0_0_green and phase0_complete".to_string(),
    );
    require(
        bool_field(claim_boundary, "additionalProperties") == Some(false),
        failures,
        "schema.claim_boundary.additionalProperties must be false".to_string(),
    );

    for field in ["build_targets", "test_targets"] {
        let target_field = object_field(properties, field).and_then(Json::as_object);
        let target_field_empty = BTreeMap::new();
        let target_field = target_field.unwrap_or(&target_field_empty);
        require(
            object_field(target_field, "minItems").and_then(Json::as_i64) == Some(1),
            failures,
            format!("schema.{field}.minItems must be 1"),
        );
    }

    SchemaSummary {
        required_fields: required,
        inventory_source_values: source_enum.into_iter().collect(),
        expected_verdict_values: verdict_enum.into_iter().collect(),
    }
}

fn inventory_fixture_shape_failures(fixture: &BTreeMap<String, Json>) -> Vec<String> {
    let mut failures = Vec::new();
    let required_fields = set_from(&REQUIRED_INVENTORY_FIELDS);
    let unknown = fixture
        .keys()
        .filter(|key| !required_fields.contains(*key))
        .cloned()
        .collect::<Vec<_>>();
    if !unknown.is_empty() {
        failures.push(format!("unexpected top-level fields: {unknown:?}"));
    }
    if !string_field(fixture, "candidate_sha").is_some_and(|value| is_hex_sha(&value)) {
        failures.push("candidate_sha must be a 40-character hexadecimal SHA".to_string());
    }
    failures
}

pub fn inventory_fixture_violations(fixture: &Json) -> Vec<String> {
    let Some(fixture) = fixture.as_object() else {
        return vec!["missing_or_malformed_trusted_target_inventory".to_string()];
    };
    let mut violations = Vec::new();
    if string_field(fixture, "inventory_source").as_deref() != Some(TRUSTED_SOURCE) {
        violations.push("target_inventory_not_trusted".to_string());
    }
    if bool_field(fixture, "captured_before_candidate_checkout") != Some(true)
        || bool_field(fixture, "candidate_checkout_after_inventory") != Some(true)
    {
        violations.push("inventory_not_captured_before_candidate_checkout".to_string());
    }
    if bool_field(fixture, "no_candidate_authored_discovery") != Some(true) {
        violations.push("candidate_can_author_target_inventory".to_string());
    }

    let build_targets = target_list(object_field(fixture, "build_targets"));
    let test_targets = target_list(object_field(fixture, "test_targets"));
    if build_targets.is_empty() || test_targets.is_empty() {
        violations.push("empty_required_targets".to_string());
    }
    if !malformed_targets(&[build_targets, test_targets].concat()).is_empty() {
        violations.push("malformed_buck2_target".to_string());
    }

    let claim_boundary = object_field(fixture, "claim_boundary").and_then(Json::as_object);
    let claim_boundary_empty = BTreeMap::new();
    let claim_boundary = claim_boundary.unwrap_or(&claim_boundary_empty);
    if bool_field(claim_boundary, "p0_0_green") != Some(false)
        || bool_field(claim_boundary, "phase0_complete") != Some(false)
    {
        violations.push("green_claim_boundary_without_live_authority".to_string());
    }
    sorted_unique(violations)
}

fn validate_fixture(
    fixture: &Json,
    path: &str,
    expected_verdict: &str,
    failures: &mut Vec<String>,
) -> FixtureValidation {
    let Some(object) = fixture.as_object() else {
        failures.push(format!("{path}: expected JSON object"));
        return FixtureValidation {
            fixture_id: path.to_string(),
            expected_verdict: expected_verdict.to_string(),
            shape_failures: vec!["expected JSON object".to_string()],
            observed_trusted_target_violations: vec![
                "missing_or_malformed_trusted_target_inventory".to_string(),
            ],
            malformed_targets: Vec::new(),
            trusted_inventory: false,
        };
    };
    let fixture_id = string_field(object, "fixture_id").unwrap_or_else(|| path.to_string());
    require(
        string_field(object, "expected_verdict").as_deref() == Some(expected_verdict),
        failures,
        format!("{fixture_id}: expected_verdict must be {expected_verdict}"),
    );
    let shape_failures = inventory_fixture_shape_failures(object);
    for shape_failure in &shape_failures {
        failures.push(format!("{fixture_id}: {shape_failure}"));
    }
    let observed_violations = inventory_fixture_violations(fixture);
    let expected_violations = string_list(object_field(object, "expected_violations"))
        .into_iter()
        .collect::<BTreeSet<_>>();
    if expected_verdict == "GREEN" {
        require(
            observed_violations.is_empty(),
            failures,
            format!(
                "{fixture_id}: GOOD trusted-target inventory has violations {observed_violations:?}"
            ),
        );
        require(
            expected_violations.is_empty(),
            failures,
            format!("{fixture_id}: GOOD fixture must not list expected violations"),
        );
    } else {
        require(
            !observed_violations.is_empty(),
            failures,
            format!("{fixture_id}: RED trusted-target inventory must produce violations"),
        );
        require(
            set_from(&TRUSTED_TARGET_VIOLATIONS).is_subset(&expected_violations),
            failures,
            format!(
                "{fixture_id}: RED fixture expected_violations must include all trusted-target violation classes"
            ),
        );
        let observed_set = observed_violations.iter().cloned().collect::<BTreeSet<_>>();
        require(
            observed_set.is_subset(&expected_violations),
            failures,
            format!(
                "{fixture_id}: observed trusted-target violations not listed in expected_violations"
            ),
        );
    }
    let build_targets = target_list(object_field(object, "build_targets"));
    let test_targets = target_list(object_field(object, "test_targets"));
    FixtureValidation {
        fixture_id,
        expected_verdict: expected_verdict.to_string(),
        shape_failures: shape_failures.clone(),
        observed_trusted_target_violations: observed_violations.clone(),
        malformed_targets: malformed_targets(&[build_targets, test_targets].concat()),
        trusted_inventory: shape_failures.is_empty() && observed_violations.is_empty(),
    }
}

pub fn evaluate_sources(
    schema: &Json,
    good_fixture: &Json,
    bad_fixture: &Json,
    schema_path: &str,
) -> Report {
    let mut failures = Vec::new();
    let schema_summary = validate_schema(schema, &mut failures);
    let fixture_results = vec![
        validate_fixture(good_fixture, DEFAULT_GOOD_FIXTURE, "GREEN", &mut failures),
        validate_fixture(bad_fixture, DEFAULT_BAD_FIXTURE, "RED", &mut failures),
    ];
    let local_fixture_contract_proven = failures.is_empty();
    Report {
        authority_boundary: AUTHORITY_BOUNDARY.to_string(),
        schema: schema_path.to_string(),
        required_inventory_fields: REQUIRED_INVENTORY_FIELDS
            .iter()
            .map(|field| (*field).to_string())
            .collect(),
        required_trusted_target_violations: TRUSTED_TARGET_VIOLATIONS
            .iter()
            .map(|violation| (*violation).to_string())
            .collect(),
        schema_summary,
        fixture_results,
        local_fixture_contract_proven,
        candidate_pr_bytes_are_data_only_locally_proven: local_fixture_contract_proven,
        trusted_target_inventory_live_authority_proven: false,
        trusted_controller_inventory_live: false,
        live_required_context_execution_proven: false,
        p0_0_green: false,
        phase0_complete: false,
        verdict: if failures.is_empty() { "PASS" } else { "FAIL" }.to_string(),
        failures,
    }
}

pub fn evaluate_paths(
    schema_path: &str,
    good_path: &str,
    bad_path: &str,
) -> Result<Report, String> {
    let schema = load_json(schema_path)?;
    let good = load_json(good_path)?;
    let bad = load_json(bad_path)?;
    Ok(evaluate_sources(&schema, &good, &bad, schema_path))
}

pub fn to_json(report: &Report) -> String {
    format!(
        concat!(
            "{{",
            "\"authority_boundary\":{},",
            "\"candidate_pr_bytes_are_data_only_locally_proven\":{},",
            "\"failures\":{},",
            "\"fixture_results\":{},",
            "\"live_required_context_execution_proven\":false,",
            "\"local_fixture_contract_proven\":{},",
            "\"p0_0_green\":false,",
            "\"phase0_complete\":false,",
            "\"required_inventory_fields\":{},",
            "\"required_trusted_target_violations\":{},",
            "\"schema\":{},",
            "\"schema_summary\":{},",
            "\"trusted_controller_inventory_live\":false,",
            "\"trusted_target_inventory_live_authority_proven\":false,",
            "\"verdict\":{}",
            "}}"
        ),
        json_string(&report.authority_boundary),
        bool_json(report.candidate_pr_bytes_are_data_only_locally_proven),
        string_array_json(&report.failures),
        fixture_results_json(&report.fixture_results),
        bool_json(report.local_fixture_contract_proven),
        string_array_json(&report.required_inventory_fields),
        string_array_json(&report.required_trusted_target_violations),
        json_string(&report.schema),
        schema_summary_json(&report.schema_summary),
        json_string(&report.verdict),
    )
}

fn schema_summary_json(summary: &SchemaSummary) -> String {
    format!(
        concat!(
            "{{",
            "\"expected_verdict_values\":{},",
            "\"inventory_source_values\":{},",
            "\"required_fields\":{}",
            "}}"
        ),
        string_array_json(&summary.expected_verdict_values),
        string_array_json(&summary.inventory_source_values),
        string_array_json(&summary.required_fields),
    )
}

fn fixture_results_json(results: &[FixtureValidation]) -> String {
    format!(
        "[{}]",
        results
            .iter()
            .map(fixture_validation_json)
            .collect::<Vec<_>>()
            .join(",")
    )
}

fn fixture_validation_json(result: &FixtureValidation) -> String {
    format!(
        concat!(
            "{{",
            "\"expected_verdict\":{},",
            "\"fixture_id\":{},",
            "\"malformed_targets\":{},",
            "\"observed_trusted_target_violations\":{},",
            "\"shape_failures\":{},",
            "\"trusted_inventory\":{}",
            "}}"
        ),
        json_string(&result.expected_verdict),
        json_string(&result.fixture_id),
        string_array_json(&result.malformed_targets),
        string_array_json(&result.observed_trusted_target_violations),
        string_array_json(&result.shape_failures),
        bool_json(result.trusted_inventory),
    )
}

fn string_array_json(values: &[String]) -> String {
    format!(
        "[{}]",
        values
            .iter()
            .map(|value| json_string(value))
            .collect::<Vec<_>>()
            .join(",")
    )
}

fn bool_json(value: bool) -> &'static str {
    if value { "true" } else { "false" }
}

fn json_string(value: &str) -> String {
    let mut out = String::from("\"");
    for ch in value.chars() {
        match ch {
            '\"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            other => out.push(other),
        }
    }
    out.push('\"');
    out
}

fn main() {
    let mut schema = DEFAULT_SCHEMA.to_string();
    let mut good_fixture = DEFAULT_GOOD_FIXTURE.to_string();
    let mut bad_fixture = DEFAULT_BAD_FIXTURE.to_string();
    let mut emit_json = false;
    let mut args = env::args().skip(1);
    while let Some(arg) = args.next() {
        match arg.as_str() {
            "--schema" => schema = args.next().expect("--schema requires a value"),
            "--good-fixture" => {
                good_fixture = args.next().expect("--good-fixture requires a value")
            }
            "--bad-fixture" => bad_fixture = args.next().expect("--bad-fixture requires a value"),
            "--json" => emit_json = true,
            other => panic!("unknown argument {other}"),
        }
    }
    let report = evaluate_paths(&schema, &good_fixture, &bad_fixture)
        .unwrap_or_else(|error| panic!("{error}"));
    let rendered = to_json(&report);
    if emit_json || report.verdict == "PASS" {
        println!("{rendered}");
    } else {
        eprintln!("{rendered}");
    }
    if report.verdict != "PASS" {
        std::process::exit(1);
    }
}
